### PR TITLE
Monitor a memory address until it reaches or is no longer a certain value in Drv X 

### PR DESCRIPTION
**Pull request by ywwu928**
# Description

## PR description

Creates an API in DrvX that can monitor a memory address until the value equals a certain value.

## Checklist

- [v] The additions follow the code standards in the [developer guide](pando-rt/docs/developer.md).
- [v] New or existing tests cover these changes.
- [v] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by AdityaAtulTewari** [comment]
How does this change drvx performance?

**Comment by ywwu928** [comment]
> How does this change drvx performance?

I would say this is more about stat collection accuracy. Rather than constanly checking whether a condition is met every 1000 cycles, I monitor if the condition is met and the DrvX memory model will inform me it is met immediately instead of waiting until the nearest multiple of 1000 cycles.

As for DrvX performance, if checking the condition requires a memory read (which happens in most of the case), then this would hep as well. We don't issue a read command every 1000 cycles anymore. Whenever there is write to the same memory address that is monitored, we will check if the condition is met.

File: include/pando-lib-galois/sync/wait_group.hpp [hunk 1]
```diff
@@ -117,11 +121,28 @@ class WaitGroup {
...
+        return ready;
+      });
+    } else {
+      DrvAPI::monitor_until<int64_t>(m_count.address, static_cast<std::int64_t>(0));
+    }
+#else
+    DrvAPI::monitor_until<int64_t>(m_count.address, static_cast<std::int64_t>(0));
+#endif
+
+#endif
```
Comment (AdityaAtulTewari) [inline include/pando-lib-galois/sync/wait_group.hpp:145]: let's make an abstract api to export this behavior

File: pando-rt/include/pando-rt/execution/result_storage.hpp [hunk 2]
```diff
@@ -57,6 +61,30 @@ struct ResultStorage {
...
+      GlobalPtr<std::int32_t> readyPtr{&ready};
+      DrvAPI::monitor_until<std::int32_t>(readyPtr.address, static_cast<std::int32_t>(1));
+    }
+#else
+    GlobalPtr<std::int32_t> readyPtr{&ready};
+    DrvAPI::monitor_until<std::int32_t>(readyPtr.address, static_cast<std::int32_t>(1));
+#endif
+
+#endif
+  }
```
Comment (AdityaAtulTewari) [inline pando-rt/include/pando-rt/execution/result_storage.hpp:87]: an abstraction would really help.

**Review by AdityaAtulTewari — COMMENTED** [review]
let's build an abstraction that in prep uses the wait until, but in drivex uses the monitor api.

**Review by AdityaAtulTewari — APPROVED** [review]

**Pull request merged**